
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
        <meta name="author" content="Genomics Aotearoa & NeSI">
      
      
        <link rel="canonical" href="https://genomicsaotearoa.github.io/shell-for-bioinformatics/5_inspectmanipulate2/">
      
      <link rel="icon" href="../images/nesi_ga.png">
      <meta name="generator" content="mkdocs-1.3.0, mkdocs-material-8.3.9">
    
    
      
        <title>Inspecting and Manipulating Text Data with Unix Tools - Part 2 - Intermediate-Advanced Shell for Bioinformatics</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.1d29e8d0.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.cbb835fc.min.css">
        
      
      
    
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Mukta:300,300i,400,400i,700,700i%7CFira+Code:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Mukta";--md-code-font:"Fira Code"}</style>
      
    
    
    <script>__md_scope=new URL("..",location),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="" data-md-color-primary="" data-md-color-accent="">
  
    
    
      <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script>
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#inspecting-and-manipulating-text-data-with-unix-tools-part-2" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="Intermediate-Advanced Shell for Bioinformatics" class="md-header__button md-logo" aria-label="Intermediate-Advanced Shell for Bioinformatics" data-md-component="logo">
      
  <img src="../images/nesi_ga.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Intermediate-Advanced Shell for Bioinformatics
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Inspecting and Manipulating Text Data with Unix Tools - Part 2
            
          </span>
        </div>
      </div>
    </div>
    
      <form class="md-header__option" data-md-component="palette">
        
          
          
          <input class="md-option" data-md-color-media="(prefers-color-scheme)" data-md-color-scheme="" data-md-color-primary="" data-md-color-accent=""  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
          
            <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_3" hidden>
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m14.3 16-.7-2h-3.2l-.7 2H7.8L11 7h2l3.2 9h-1.9M20 8.69V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69m-9.15 3.96h2.3L12 9l-1.15 3.65Z"/></svg>
            </label>
          
        
          
          
          <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent=""  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_2">
          
            <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg>
            </label>
          
        
          
          
          <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="" data-md-color-accent=""  aria-label="Switch to system preference"  type="radio" name="__palette" id="__palette_3">
          
            <label class="md-header__button md-icon" title="Switch to system preference" for="__palette_2" hidden>
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg>
            </label>
          
        
      </form>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/GenomicsAotearoa/shell-for-bioinformatics.git" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.1.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GenomicsAotearoa/shell-for-bioinformatics
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="Intermediate-Advanced Shell for Bioinformatics" class="md-nav__button md-logo" aria-label="Intermediate-Advanced Shell for Bioinformatics" data-md-component="logo">
      
  <img src="../images/nesi_ga.png" alt="logo">

    </a>
    Intermediate-Advanced Shell for Bioinformatics
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/GenomicsAotearoa/shell-for-bioinformatics.git" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.1.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GenomicsAotearoa/shell-for-bioinformatics
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        Home
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../0_introduction/" class="md-nav__link">
        UNIX, Linux  & UNIX Shell
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../1_download_data/" class="md-nav__link">
        Download & Verify Data
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../2_unixshellbasics/" class="md-nav__link">
        Unix Shell Basics & Recap
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../4_inspectmanipluate/" class="md-nav__link">
        Inspecting and Manipulating Text Data with Unix Tools - Part 1
      </a>
    </li>
  

    
      
      
      

  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Inspecting and Manipulating Text Data with Unix Tools - Part 2
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        Inspecting and Manipulating Text Data with Unix Tools - Part 2
      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#aho-weinberger-kernighan-awk" class="md-nav__link">
    Aho, Weinberger, Kernighan = AWK
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#_1" class="md-nav__link">
    
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#bioawk" class="md-nav__link">
    bioawk
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../6_automate_fileprocessing_find_xargs/" class="md-nav__link">
        Automating File-Processing with find and xargs
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../puzzles/" class="md-nav__link">
        Puzzles
      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#aho-weinberger-kernighan-awk" class="md-nav__link">
    Aho, Weinberger, Kernighan = AWK
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#_1" class="md-nav__link">
    
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#bioawk" class="md-nav__link">
    bioawk
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content" data-md-component="content">
            <article class="md-content__inner md-typeset">
              
                
  <a href="https://github.com/GenomicsAotearoa/shell-for-bioinformatics.git/edit/master/docs/5_inspectmanipulate2.md" title="Edit this page" class="md-content__button md-icon">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25Z"/></svg>
  </a>



<h1 id="inspecting-and-manipulating-text-data-with-unix-tools-part-2">Inspecting and Manipulating Text Data with Unix Tools - Part 2<a class="headerlink" href="#inspecting-and-manipulating-text-data-with-unix-tools-part-2" title="Permanent link">&para;</a></h1>
<div class="admonition abstract">
<p class="admonition-title">Lesson Objectives</p>
</div>
<h2 id="aho-weinberger-kernighan-awk">Aho, Weinberger, Kernighan = AWK<a class="headerlink" href="#aho-weinberger-kernighan-awk" title="Permanent link">&para;</a></h2>
<p>Awk is a scripting language used for manipulating data and generating reports. The awk command programming language requires no compiling and allows the user to use variables, numeric functions, string functions, and logical operators. </p>
<p>Awk is a utility that enables a programmer to write tiny but effective programs in the form of statements that define text patterns that are to be searched for in each line of a document and the action that is to be taken when a match is found within a line. Awk is mostly used for pattern scanning and processing. It searches one or more files to see if they contain lines that matches with the specified patterns and then perform the associated actions. </p>
<details open="open">
<summary>WHAT CAN WE DO WITH AWK?</summary>
<ol>
<li>
<p>AWK Operations: </p>
<ul>
<li>Scans a file line by line </li>
<li>Splits each input line into fields </li>
<li>Compares input line/fields to pattern </li>
<li>Performs action(s) on matched lines </li>
</ul>
</li>
<li>
<p>Useful For: </p>
<ul>
<li>Transform data files </li>
<li>Produce formatted reports </li>
</ul>
</li>
<li>
<p>Programming Constructs: </p>
<ul>
<li>Format output lines </li>
<li>Arithmetic and string operations </li>
<li>Conditionals and loops </li>
</ul>
</li>
</ol>
</details>
<p><strong>Syntax</strong>:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a>awk options <span class="s1">&#39;selection_criteria {action}&#39;</span> input-file &gt;  output-file
</code></pre></div>
<div class="admonition options">
<p class="admonition-title">Options</p>
<p><code>-f program-file</code> : Reads the AWK program source from the file program-file, instead of from the first command line argument.</p>
<p><code>-F fs</code>            : Use fs for the input field separator</p>
</div>
<p>Default behaviour of <code>awk</code> is to print every line of data from the specified file. .i.e. mimics <code>cat</code></p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a>awk <span class="s1">&#39;{print}&#39;</span> example.bed 
</code></pre></div>
<details class="success">
<summary>Output</summary>
<div class="highlight"><pre><span></span><code><a id="__codelineno-2-1" name="__codelineno-2-1" href="#__codelineno-2-1"></a>chr1    <span class="m">26</span>  <span class="m">39</span>
<a id="__codelineno-2-2" name="__codelineno-2-2" href="#__codelineno-2-2"></a>chr1    <span class="m">32</span>  <span class="m">47</span>
<a id="__codelineno-2-3" name="__codelineno-2-3" href="#__codelineno-2-3"></a>chr3    <span class="m">11</span>  <span class="m">28</span>
<a id="__codelineno-2-4" name="__codelineno-2-4" href="#__codelineno-2-4"></a>chr1    <span class="m">40</span>  <span class="m">49</span>
<a id="__codelineno-2-5" name="__codelineno-2-5" href="#__codelineno-2-5"></a>chr3    <span class="m">16</span>  <span class="m">27</span>
<a id="__codelineno-2-6" name="__codelineno-2-6" href="#__codelineno-2-6"></a>chr1    <span class="m">9</span>   <span class="m">28</span>
<a id="__codelineno-2-7" name="__codelineno-2-7" href="#__codelineno-2-7"></a>chr2    <span class="m">35</span>  <span class="m">54</span>
<a id="__codelineno-2-8" name="__codelineno-2-8" href="#__codelineno-2-8"></a>chr1    <span class="m">10</span>  <span class="m">19</span>
</code></pre></div>
</details>
<p>Print lines which match the given pattern</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-3-1" name="__codelineno-3-1" href="#__codelineno-3-1"></a>awk <span class="s1">&#39;/chr1/ {print}&#39;</span> example.bed 
<a id="__codelineno-3-2" name="__codelineno-3-2" href="#__codelineno-3-2"></a>
<a id="__codelineno-3-3" name="__codelineno-3-3" href="#__codelineno-3-3"></a>chr1    <span class="m">26</span>  <span class="m">39</span>
<a id="__codelineno-3-4" name="__codelineno-3-4" href="#__codelineno-3-4"></a>chr1    <span class="m">32</span>  <span class="m">47</span>
<a id="__codelineno-3-5" name="__codelineno-3-5" href="#__codelineno-3-5"></a>chr1    <span class="m">40</span>  <span class="m">49</span>
<a id="__codelineno-3-6" name="__codelineno-3-6" href="#__codelineno-3-6"></a>chr1    <span class="m">9</span>   <span class="m">28</span>
<a id="__codelineno-3-7" name="__codelineno-3-7" href="#__codelineno-3-7"></a>chr1    <span class="m">10</span>  <span class="m">19</span>
</code></pre></div>
<p><code>awk</code> can be used to mimic functionality of <code>cut</code> </p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-4-1" name="__codelineno-4-1" href="#__codelineno-4-1"></a>awk <span class="s1">&#39;{print $2 &quot;\t&quot; $3}&#39;</span> example.bed 
</code></pre></div>
<details class="success">
<summary>Output</summary>
<div class="highlight"><pre><span></span><code><a id="__codelineno-5-1" name="__codelineno-5-1" href="#__codelineno-5-1"></a><span class="m">26</span>  <span class="m">39</span>
<a id="__codelineno-5-2" name="__codelineno-5-2" href="#__codelineno-5-2"></a><span class="m">32</span>  <span class="m">47</span>
<a id="__codelineno-5-3" name="__codelineno-5-3" href="#__codelineno-5-3"></a><span class="m">11</span>  <span class="m">28</span>
<a id="__codelineno-5-4" name="__codelineno-5-4" href="#__codelineno-5-4"></a><span class="m">40</span>  <span class="m">49</span>
<a id="__codelineno-5-5" name="__codelineno-5-5" href="#__codelineno-5-5"></a><span class="m">16</span>  <span class="m">27</span>
<a id="__codelineno-5-6" name="__codelineno-5-6" href="#__codelineno-5-6"></a> <span class="m">9</span>  <span class="m">28</span>
<a id="__codelineno-5-7" name="__codelineno-5-7" href="#__codelineno-5-7"></a><span class="m">35</span>  <span class="m">54</span>
<a id="__codelineno-5-8" name="__codelineno-5-8" href="#__codelineno-5-8"></a><span class="m">10</span>  <span class="m">19</span>
</code></pre></div>
</details>
<p>Here, we’re making use of Awk’s string concatenation. Two strings are concatenated if they are placed next to each other with no argument. So for each record, <code>$2"\t"$3</code> concatenates the second field, a tab character, and the third field. However, this is an instance where using <code>cut</code> is much simpler as the equivalent of above is <code>cut -f2,3 example.bed</code> </p>
<p>Let’s look at how we can incorporate simple pattern matching. Suppose we wanted to write a filter that only output lines where the length of the feature (end
position - start position) was greater than 18. Awk supports arithmetic with the standard operators + , - , * , / , % (remainder), and ^ (exponentiation). We can subtract within a pattern to calculate the length of a feature, and filter on that expression:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-6-1" name="__codelineno-6-1" href="#__codelineno-6-1"></a>awk <span class="s1">&#39;$3 - $2 &gt; 18&#39;</span> example.bed 
<a id="__codelineno-6-2" name="__codelineno-6-2" href="#__codelineno-6-2"></a>
<a id="__codelineno-6-3" name="__codelineno-6-3" href="#__codelineno-6-3"></a>chr1    <span class="m">9</span>   <span class="m">28</span>
<a id="__codelineno-6-4" name="__codelineno-6-4" href="#__codelineno-6-4"></a>chr2    <span class="m">35</span>  <span class="m">54</span>
</code></pre></div>
<hr />
<p><center></p>
<table>
<thead>
<tr>
<th align="left">Comparison</th>
<th align="left">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left"><code>a == b</code></td>
<td align="left">a is equal to b</td>
</tr>
<tr>
<td align="left"><code>a != b</code></td>
<td align="left">a is not equal to b</td>
</tr>
<tr>
<td align="left"><code>a &lt; b</code></td>
<td align="left">a is less than b</td>
</tr>
<tr>
<td align="left"><code>a &gt; b</code></td>
<td align="left">a is greater than b</td>
</tr>
<tr>
<td align="left"><code>a &lt;= b</code></td>
<td align="left">a is less than or equal to b</td>
</tr>
<tr>
<td align="left"><code>a &gt;= b</code></td>
<td align="left">a is greater than or equal to b</td>
</tr>
<tr>
<td align="left"><code>a ~ b</code></td>
<td align="left">a matches regular expression pattern b</td>
</tr>
<tr>
<td align="left"><code>a !~ b</code></td>
<td align="left">a does not match regular expression pattern b</td>
</tr>
<tr>
<td align="left"><code>a &amp;&amp; b</code></td>
<td align="left">logical a and b</td>
</tr>
<tr>
<td align="left"><code>a \|\| b</code></td>
<td align="left">logical or a and b</td>
</tr>
<tr>
<td align="left"><code>!a</code></td>
<td align="left">not a (logical negation)</td>
</tr>
</tbody>
</table>
<h2 id="_1"></center><a class="headerlink" href="#_1" title="Permanent link">&para;</a></h2>
<p>We can also chain patterns, by using logical operators <code>&amp;&amp;</code> (AND), <code>||</code> (OR), and <code>!</code> (NOT). For example, if we wanted all lines on chromosome 1 with a length greater than 10:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-7-1" name="__codelineno-7-1" href="#__codelineno-7-1"></a>awk <span class="s1">&#39;$1 ~ /chr1/ &amp;&amp; $3 - $2 &gt; 10&#39;</span> example.bed 
<a id="__codelineno-7-2" name="__codelineno-7-2" href="#__codelineno-7-2"></a>
<a id="__codelineno-7-3" name="__codelineno-7-3" href="#__codelineno-7-3"></a>chr1    <span class="m">26</span>  <span class="m">39</span>
<a id="__codelineno-7-4" name="__codelineno-7-4" href="#__codelineno-7-4"></a>chr1    <span class="m">32</span>  <span class="m">47</span>
<a id="__codelineno-7-5" name="__codelineno-7-5" href="#__codelineno-7-5"></a>chr1    <span class="m">9</span>   <span class="m">28</span>
</code></pre></div>
<div class="admonition info">
<ul>
<li>First pattern, <code>$1 ~ /chr1</code> specifies the regular expression. (All Regular expressions are in slashes.)  We are mating the first field, <code>$1</code> against the regular expression <code>chr1</code>. </li>
<li>Tilde <code>~</code> means <strong>match</strong>.</li>
<li>To do the inverse of <strong>match</strong>, we can use <code>!~</code> OR <code>!($1 ~ /chr1/</code></li>
</ul>
</div>
<div class="admonition note">
<p class="admonition-title">Built-In Variables and special patterns In Awk</p>
<ul>
<li>
<p>Awk’s built-in variables include the field variables <code>$1</code>, <code>$2</code>, <code>$3</code>, and so on (<code>$0</code> is the entire line) — that break a line of text into individual words or pieces called fields. </p>
<ul>
<li><code>NR</code>: keeps a current count of the number of input records. Remember that records are usually lines. Awk command performs the pattern/action statements once for each record in a file. </li>
<li><code>NF</code>: keeps a count of the number of fields within the current input record. </li>
<li><code>FS</code>: contains the field separator character which is used to divide fields on the input line. The default is “white space”, meaning space and tab characters. FS can be reassigned to another character (typically in BEGIN) to change the field separator. </li>
<li><code>RS</code>: stores the current record separator character. Since, by default, an input line is the input record, the default record separator character is a newline. </li>
<li><code>OFS</code>: stores the output field separator, which separates the fields when Awk prints them. The default is a blank space. Whenever print has several parameters separated with commas, it will print the value of OFS in between each parameter. </li>
<li><code>ORS</code>: stores the output record separator, which separates the output lines when Awk prints them. The default is a newline character. print automatically outputs the contents of ORS at the end of whatever it is given to print. </li>
</ul>
</li>
<li>
<p>Also, there are two special patterns <code>BEGIN</code> &amp; <code>END</code></p>
<ul>
<li><code>BEGIN</code> - specifies what to do before the first record is read in. Useful to initialise and set up variables</li>
<li><code>END</code> - what to do after the last record's processing is complete. Useful to print data summaries ad the end of file processing</li>
</ul>
</li>
</ul>
</div>
<p><strong>Examples</strong></p>
<ul>
<li>We can use <code>NR</code> to extract ranges of lines, too; for example, if we wanted to extract all lines between 3 and 5 (inclusive):</li>
</ul>
<div class="highlight"><pre><span></span><code><a id="__codelineno-8-1" name="__codelineno-8-1" href="#__codelineno-8-1"></a>awk <span class="s1">&#39;NR &gt;= 3 &amp;&amp; NR &lt;=5&#39;</span> example.bed 
<a id="__codelineno-8-2" name="__codelineno-8-2" href="#__codelineno-8-2"></a>
<a id="__codelineno-8-3" name="__codelineno-8-3" href="#__codelineno-8-3"></a>    chr3    <span class="m">11</span>  <span class="m">28</span>
<a id="__codelineno-8-4" name="__codelineno-8-4" href="#__codelineno-8-4"></a>    chr1    <span class="m">40</span>  <span class="m">49</span>
<a id="__codelineno-8-5" name="__codelineno-8-5" href="#__codelineno-8-5"></a>    chr3    <span class="m">16</span>  <span class="m">27</span>
</code></pre></div>
<ul>
<li>suppose we wanted to calculate the mean feature length in example.bed. We would have to take the sum feature lengths, and then divide by the total number of records. We can do this with:</li>
</ul>
<div class="highlight"><pre><span></span><code><a id="__codelineno-9-1" name="__codelineno-9-1" href="#__codelineno-9-1"></a>awk <span class="s1">&#39;BEGIN{s = 0}; {s += ($3-$2)}; END{ print &quot;mean: &quot; s/NR};&#39;</span> example.bed 
<a id="__codelineno-9-2" name="__codelineno-9-2" href="#__codelineno-9-2"></a>
<a id="__codelineno-9-3" name="__codelineno-9-3" href="#__codelineno-9-3"></a>  mean: <span class="m">14</span>
</code></pre></div>
<div class="admonition info">
<p class="admonition-title">Info</p>
<p>In this example, we’ve initialized a variable <code>s</code> to <strong>0</strong> in <code>BEGIN</code> (variables you define do not need a dollar sign). Then, for each record we increment s by the length of the feature. At the end of the records, we print this sum <code>s</code> divided by the number of records <code>NR</code> , giving the mean.</p>
</div>
<ul>
<li><code>awk</code> makes it easy to convert between bioinformatics files like BED and GTF. For example, we could generate a three-column BED file from <strong><em>Mus_muscu‐lus.GRCm38.75_chr1.gtf</em></strong> as follows:</li>
</ul>
<div class="highlight"><pre><span></span><code><a id="__codelineno-10-1" name="__codelineno-10-1" href="#__codelineno-10-1"></a>awk <span class="s1">&#39;!/^#/ { print $1 &quot;\t&quot; $4-1 &quot;\t&quot; $5}&#39;</span> Mus_musculus.GRCm38.75_chr1.gtf <span class="p">|</span> head -n <span class="m">3</span>
<a id="__codelineno-10-2" name="__codelineno-10-2" href="#__codelineno-10-2"></a>
<a id="__codelineno-10-3" name="__codelineno-10-3" href="#__codelineno-10-3"></a><span class="m">1</span>   <span class="m">3054232</span> <span class="m">3054733</span>
<a id="__codelineno-10-4" name="__codelineno-10-4" href="#__codelineno-10-4"></a><span class="m">1</span>   <span class="m">3054232</span> <span class="m">3054733</span>
<a id="__codelineno-10-5" name="__codelineno-10-5" href="#__codelineno-10-5"></a><span class="m">1</span>   <span class="m">3054232</span> <span class="m">3054733</span>
</code></pre></div>
<ul>
<li><code>awk</code> also has a very useful data structure known as an associative array. Associative arrays behave like Python’s dictionaries or hashes in other languages. We can create an associative array by simply assigning a value to a key. For example, suppose we wanted to count the number of features (third column) belonging to the gene “Lypla1.” We could do this by incrementing their values in an associative array:</li>
</ul>
<div class="highlight"><pre><span></span><code><a id="__codelineno-11-1" name="__codelineno-11-1" href="#__codelineno-11-1"></a>awk <span class="s1">&#39;/Lypla1/ {feature[$3] += 1}; END {for (k in feature) print k &quot;\t&quot; feature[k]}&#39;</span> Mus_musculus.GRCm38.75_chr1.gtf 
<a id="__codelineno-11-2" name="__codelineno-11-2" href="#__codelineno-11-2"></a>
<a id="__codelineno-11-3" name="__codelineno-11-3" href="#__codelineno-11-3"></a>CDS <span class="m">56</span>
<a id="__codelineno-11-4" name="__codelineno-11-4" href="#__codelineno-11-4"></a>transcript  <span class="m">9</span>
<a id="__codelineno-11-5" name="__codelineno-11-5" href="#__codelineno-11-5"></a>start_codon <span class="m">5</span>
<a id="__codelineno-11-6" name="__codelineno-11-6" href="#__codelineno-11-6"></a>gene    <span class="m">1</span>
<a id="__codelineno-11-7" name="__codelineno-11-7" href="#__codelineno-11-7"></a>exon    <span class="m">69</span>
<a id="__codelineno-11-8" name="__codelineno-11-8" href="#__codelineno-11-8"></a>UTR <span class="m">24</span>
</code></pre></div>
<h2 id="bioawk"><code>bioawk</code><a class="headerlink" href="#bioawk" title="Permanent link">&para;</a></h2>
<p><code>bioawk</code> is an extension of <code>awk</code>, adding the support of several common biological data formats, including optionally gzip'ed BED, GFF, SAM, VCF, FASTA/Q and TAB-delimited formats with column names. It also adds a few built-in functions and an command line option to use TAB as the input/output delimiter. When the new functionality is not used, <code>bioawk</code> is intended to behave exactly the same as the original BWK awk.</p>
<p>The original <code>awk</code> requires a YACC-compatible parser generator (e.g. Byacc or Bison). <code>bioawk</code> further depends on <a href="http://zlib.net/">zlib</a> so as to work with gzip'd files.</p>
<details>
<summary>YACC</summary>
<p>A parser generator is a program that takes as input a specification of a syntax, and produces as output a procedure for recognizing that language. Historically, they are also called compiler-compilers.
  YACC (yet another compiler-compiler) is an LALR(LookAhead, Left-to-right, Rightmost derivation producer with 1 lookahead token) parser generator. YACC was originally designed for being complemented by <strong>Lex</strong>.</p>
<ul>
<li><strong>Lex</strong> (A Lexical Analyzer Generator) helps write programs whose control flow is directed by instances of regular expressions in the input stream. It is well suited for editor-script type transformations and for segmenting input in preparation for a parsing routine. </li>
</ul>
</details>
<div class="admonition info">
<p class="admonition-title"><code>bioawk</code> features</p>
<ul>
<li>It can automatically recognize some popular formats and will parse different features associated with those formats. The format option is passed to <code>bioawk</code> using <code>-c arg</code> flag. Here <code>arg</code> can be bed, sam, vcf, gff or fastx (for both fastq and FASTA). It can also deal with other types of table formats using the <code>-c header</code> option. When <code>header</code> is specified, the field names will used for variable names, thus greatly expanding the utility.`</li>
<li>There are several builtin functions (other than the standard <code>awk</code> built-ins), that are specific to biological file formats. When a format is read with <code>bioawk</code>, the fields get automatically parsed. You can apply several functions on these variables to get the desired output. Let’s say, we read fasta format, now we have <code>$name</code> and <code>$seq</code> that holds sequence name and sequence respectively. You can use the <code>print</code> function (<code>awk</code> builtin) to print <code>$name</code> and <code>$seq</code>. You can also use <code>bioawk</code> built-in with the <code>print</code> function to get length, reverse complement etc by just using <code>'{print length($seq)}'</code>. Other functions include <code>reverse</code>, <code>revcomp</code>, <code>trimq</code>, <code>and</code>, <code>or</code>, <code>xor</code> etc.</li>
</ul>
</div>
<details>
<summary>Variables for each format</summary>
<p>For the <code>-c</code> you can either specify bed, sam, vcf, gff, fastx or header. <code>bioawk</code> will parse these variables for the respective format. If <code>-c</code> header is specified, the field names (first line) will be used as variables (spaces and special characters will be changed to under_score)</p>
<p><center></p>
<table>
<thead>
<tr>
<th align="left"><strong>bed</strong></th>
<th align="left"><strong>sam</strong></th>
<th align="left"><strong>vcf</strong></th>
<th align="left"><strong>gff</strong></th>
<th align="left"><strong>fastx</strong></th>
</tr>
</thead>
<tbody>
<tr>
<td align="left">chrom</td>
<td align="left">qname</td>
<td align="left">chrom</td>
<td align="left">seqname</td>
<td align="left">name</td>
</tr>
<tr>
<td align="left">start</td>
<td align="left">flag</td>
<td align="left">pos</td>
<td align="left">source</td>
<td align="left">seq</td>
</tr>
<tr>
<td align="left">end</td>
<td align="left">rname</td>
<td align="left">id</td>
<td align="left">feature</td>
<td align="left">qual</td>
</tr>
<tr>
<td align="left">name</td>
<td align="left">pos</td>
<td align="left">ref</td>
<td align="left">start</td>
<td align="left">comment</td>
</tr>
<tr>
<td align="left">score</td>
<td align="left">mapq</td>
<td align="left">alt</td>
<td align="left">end</td>
<td align="left"></td>
</tr>
<tr>
<td align="left">strand</td>
<td align="left">cigar</td>
<td align="left">qual</td>
<td align="left">score</td>
<td align="left"></td>
</tr>
<tr>
<td align="left">thickstart</td>
<td align="left">rnext</td>
<td align="left">filter</td>
<td align="left">filter</td>
<td align="left"></td>
</tr>
<tr>
<td align="left">thickend</td>
<td align="left">pnext</td>
<td align="left">info</td>
<td align="left">strand</td>
<td align="left"></td>
</tr>
<tr>
<td align="left">rgb</td>
<td align="left">tlen</td>
<td align="left">group</td>
<td align="left"></td>
<td align="left"></td>
</tr>
<tr>
<td align="left">blockcount</td>
<td align="left">seq</td>
<td align="left">attribute</td>
<td align="left"></td>
<td align="left"></td>
</tr>
<tr>
<td align="left">blocksizes</td>
<td align="left">qual</td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
</tr>
<tr>
<td align="left">blockstarts</td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
<td align="left"></td>
</tr>
</tbody>
</table>
<p></center></p>
</details>
<hr />
<p align="center"><b><a class="btn" href="https://genomicsaotearoa.github.io/shell-for-bioinformatics/" style="background: var(--bs-dark);font-weight:bold">Back to homepage</a></b></p>

              
            </article>
            
          </div>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
    
    <nav class="md-footer__inner md-grid" aria-label="Footer" >
      
        
        <a href="../4_inspectmanipluate/" class="md-footer__link md-footer__link--prev" aria-label="Previous: Inspecting and Manipulating Text Data with Unix Tools - Part 1" rel="prev">
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
          </div>
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Previous
              </span>
              Inspecting and Manipulating Text Data with Unix Tools - Part 1
            </div>
          </div>
        </a>
      
      
        
        <a href="../6_automate_fileprocessing_find_xargs/" class="md-footer__link md-footer__link--next" aria-label="Next: Automating File-Processing with find and xargs" rel="next">
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Next
              </span>
              Automating File-Processing with find and xargs
            </div>
          </div>
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg>
          </div>
        </a>
      
    </nav>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    <script id="__config" type="application/json">{"base": "..", "features": ["navigation.tracking", "navigation.sections", "navigation.expand"], "search": "../assets/javascripts/workers/search.b97dbffb.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version.title": "Select version"}}</script>
    
    
      <script src="../assets/javascripts/bundle.6c7ad80a.min.js"></script>
      
    
  </body>
</html>